### PR TITLE
feat: allow custom nonce in OIDAuthorizationRequest

### PR DESCRIPTION
**Pull request by vonovak**
this PR is in response to this PR which has been opened ~ year ago and appears to be in a limbo

https://github.com/google/GoogleSignIn-iOS/pull/244

The PR did get a review from @mdmathias who pointed out one issue [here](https://github.com/google/GoogleSignIn-iOS/pull/244#discussion_r1213748728) that seems to prevent the PR from getting merged. 

There is a substantial community interest to get the PR merged and it appears there is willingness on the google side as well.

I'm attempting to unblock the PR by adding the initializer that is being asked for in the review comment (linked above).

Thank you in advance for your review

### CODE REVIEW

**Comment by vonovak** [comment]
I signed the contributor agreement. According to https://github.com/openid/AppAuth-iOS/blob/master/CONTRIBUTING.md there's two to sign, but I only really saw one. 

Hope somebody's going to review this :D

File: Source/AppAuthCore/OIDAuthorizationRequest.h [hunk 1]
```diff
@@ -159,6 +159,29 @@ extern NSString *const OIDOAuthorizationRequestCodeChallengeMethodS256;
...
      additionalParameters:(nullable NSDictionary<NSString *, NSString *> *)additionalParameters;
 
+/*! @brief Creates an authorization request with custom nonce, a secure @c state,
+        and PKCE with S256 as the @c code_challenge_method.
+    @param configuration The service's configuration.
+    @param clientID The client identifier.
+    @param scopes An array of scopes to combine into a single scope string per the OAuth2 spec.
+    @param redirectURL The client's redirect URI.
+    @param responseType The expected response type.
+    @param nonce String value used to associate a Client session with an ID Token. Can be set to nil
```
Comment (mdmathias) [inline Source/AppAuthCore/OIDAuthorizationRequest.h:169]: Should the `nonce` parameter below be marked `nullable`?

File: Source/AppAuthCore/OIDAuthorizationRequest.m [hunk 2]
```diff
@@ -202,6 +202,32 @@ - (instancetype)initWithConfiguration:(OIDServiceConfiguration *)configuration
                 additionalParameters:additionalParameters];
 }
 
+- (instancetype)
+    initWithConfiguration:(OIDServiceConfiguration *)configuration
+                 clientId:(NSString *)clientID
+                   scopes:(nullable NSArray<NSString *> *)scopes
+              redirectURL:(NSURL *)redirectURL
+             responseType:(NSString *)responseType
+                    nonce:(NSString *)nonce
```
Comment (mdmathias) [inline Source/AppAuthCore/OIDAuthorizationRequest.m:211]: Mark as `nullable`.

File: UnitTests/OIDAuthorizationRequestTests.m [hunk 3]
```diff
@@ -223,6 +223,29 @@ - (void)testScopeInitializerWithManyScopesAndNoClientSecret {
...
+  OIDAuthorizationRequest *request =
+      [[OIDAuthorizationRequest alloc] initWithConfiguration:configuration
+                                                    clientId:kTestClientID
+                                                      scopes:@[]
+                                                 redirectURL:[NSURL URLWithString:kTestRedirectURL]
+                                                responseType:OIDResponseTypeCode
+                                                       nonce:kTestNonce
+                                        additionalParameters:nil];
+
+  XCTAssertEqualObjects(request.nonce, kTestNonce, @"");
```
Comment (mdmathias) [inline UnitTests/OIDAuthorizationRequestTests.m:240]: I recognize that you're following the pre-existing style in this file by adding an empty description of the failure in each of these asserts, but that is an odd practice IMHO. Let's remove these as they don't add anything and consider this as a first step in migrating away from that practice.

**Review by mdmathias — CHANGES_REQUESTED** [review]

File: UnitTests/OIDAuthorizationRequestTests.m (hunk 3 above)
Comment (vonovak) [reply UnitTests/OIDAuthorizationRequestTests.m]: I updated the failure message, this is what it looks like now:

Screenshot 2023-10-04 at 21 03 26


Please let me know if this is good, or if I should add that to the other assertions as well.

Thank you!

Comment (mdmathias) [reply UnitTests/OIDAuthorizationRequestTests.m]: You can either add meaningful descriptions (like you have in your screenshot) or remove the parameter completely. Up to you!

**Comment by mdmathias** [comment]
@vonovak I will be OOO for about 1.5 weeks starting tomorrow. I promise to take a look when I am back. Thanks!

File: UnitTests/OIDAuthorizationRequestTests.m (hunk 3 above)
Comment (mdmathias) [reply UnitTests/OIDAuthorizationRequestTests.m]: I'm sorry -- it seems like we are having a miscommunication. I'm saying to remove the argument you supply to the third parameter completely. This,

`XCTAssertEqualObjects(request.nonce, kTestNonce, @"");`

Should become

`XCTAssertEqualObjects(request.nonce, kTestNonce);`

The same very every assert in this test that you're adding. As I mentioned before, I recognize that the style is different in the other tests that were already in place in this file, but let's consider your new test as a step in the new direction.

Comment (vonovak) [reply UnitTests/OIDAuthorizationRequestTests.m]: removed the 3rd param @mdmathias 👍

File: UnitTests/OIDAuthorizationRequestTests.m [hunk 4]
```diff
@@ -223,6 +223,29 @@ - (void)testScopeInitializerWithManyScopesAndNoClientSecret {
...
+                                                 redirectURL:[NSURL URLWithString:kTestRedirectURL]
+                                                responseType:OIDResponseTypeCode
+                                                       nonce:kTestNonce
+                                        additionalParameters:nil];
+
+  XCTAssertEqualObjects(request.nonce, kTestNonce);
+  XCTAssertEqualObjects(request.responseType, @"code");
+  XCTAssertEqualObjects(request.scope, @"");
+  XCTAssertEqualObjects(request.clientID, kTestClientID);
+  XCTAssertEqualObjects(request.clientSecret, nil);
```
Comment (mdmathias) [inline UnitTests/OIDAuthorizationRequestTests.m:244]: Why not use [XCTAssertNil](https://developer.apple.com/documentation/xctest/xctassertnil)?

**Review by mdmathias — CHANGES_REQUESTED** [review]
Sorry! Just one last suggestion to make an assert more clear. Ready to go after this.

File: UnitTests/OIDAuthorizationRequestTests.m (hunk 4 above)
Comment (vonovak) [reply UnitTests/OIDAuthorizationRequestTests.m]: used it 👍

**Review by mdmathias — APPROVED** [review]

**Comment by vonovak** [comment]
thanks for review and merge! 🙂

**Comment by mdmathias** [comment]
Absolutely, thanks for the contribution. :D

**Comment by vonovak** [comment]
@mdmathias would you please share an ETA for the next release? I'd like to integrate this change into https://github.com/google/GoogleSignIn-iOS

thank you

**Comment by mdmathias** [comment]
We are currently working on adding a privacy manifest for this library, and the next release will be out before Spring 2024.

**Pull request merged**